src/charm.py

```diff
@@ -25,6 +25,45 @@
 SCRAPE_RELATION_NAME = "metrics-endpoint"
 
 
+def key_value_pair_string_to_dict(key_value_pair: str) -> dict:
+    """Transform a comma-separated key-value pairs into a dict."""
+    result = {}
+
+    for pair in key_value_pair.split(","):
+        pair = pair.strip()
+        if not pair:
+            continue
+
+        if ":" in pair:
+            sep = ":"
+        elif "=" in pair:
+            sep = "="
+        else:
+            logger.error("Invalid pair without separator ':' or '=': '%s'", pair)
+            continue
+
+        key, value = map(str.strip, pair.split(sep, 1))
+
+        if not key:
+            logger.error("Empty key in pair: '%s'", pair)
+            continue
+        if not value:
+            logger.error("Empty value in pair: '%s'", pair)
+            continue
+
+        result[key] = value
+
+    return result
+
+def inject_extra_labels_to_alert_rules(rules: dict, extra_alert_labels: dict) -> dict:
+    """Inject extra alert labels into alert labels."""
+    for _, item in rules.items():
+        for groups in item.get("groups", []):
+            for rule in groups.get("rules", []):
+                rule.setdefault("labels", {}).update(extra_alert_labels)
+    return rules
```

Comment:
The return here is a bit confusing because the function mutates the "by ref" arg `rules`.
Seems like we should operate on `rules.copy()` or not return anything?